add system locking chapter (#278)
uyuni-project/uyuni-docs: modules/client-configuration/pages/system-locking.adoc

[[system=locking]]
= System Locking

System locks are used to prevent actions from occurring on a client.
For example, a system lock will prevent a client from being updated or restarted.
This is useful for clients running production software, or to prevent accidental changes.
You can disable the system lock when you are ready to perform actions.

System locks are implemented differently on traditional and Salt clients.

== System Locks on Traditional Clients

> When a traditional client is locked, no actions can be scheduled, and a padlock icon is displayed next to the name of the client in the menu:System[System List].

Reviewer: Remote commands can still be issued an executed even when a traditional system is locked

To lock a traditional client using the {productname} {webui}, navigate to the [guimenu]``System Details`` page for the client you want to lock.
Under [guimenu]``Lock Status``, click btn:[Lock this system].

The client will remain locked until you click btn:[Unlock this system].

//Something about requiring admin rights here maybe? --LKB 20200514

[IMPORTANT]
====
Automated patch updates can still be applied to system locked traditional clients.
To stop automated patch updates, navigate to the [guimenu]``System Details`` page for the client, and on the [guimenu]``Properties`` tab, uncheck [guimenu]``Auto Patch Update``.
====

// All these webui things need to be checked, ref server is down right now. --LKB 20200514

== System Locks on Salt Clients

> When a Salt client is locked, or put into blackout mode, no actions can be scheduled, and a yellow banner is displayed on the [guimenu]``System Details`` page.

Reviewer: ```suggestion
When a Salt client is locked, or put into [Salt blackout mode](https://docs.saltstack.com/en/latest/topics/blackout/index.html), no actions can be scheduled, and a yellow banner is displayed on the [guimenu]``System Details`` page. Additionally, any Salt execution command is disabled (except for the ones explicitly allowed in the system lock formula configuration).
```

In this mode, actions can be scheduled for the locked client using the {webui} or the API, but the actions will fail.

[NOTE]
====
The locking mechanism is not available for Salt SSH clients.
====

To lock a Salt client using the {productname} {webui}, navigate to the [guimenu]``System Details`` page for the client you want to lock.

> Navigate to the [guimenu]``Formulas`` tab, enable the system lock formula, and apply the highstate.

Reviewer: ```suggestion
Navigate to the [guimenu]``Formulas`` tab, enable the system lock formula, and tick the checkbox "Lock system". Finally, save the formula (applying the highstate is not required for this formula).
```

Author: ![image](https://user-images.githubusercontent.com/3914967/82003886-da549280-96a4-11ea-9f0e-219dd582f930.png)
It actually tells you to apply the highstate in the UI :confused:

Reviewer: Exactly:
![image](https://user-images.githubusercontent.com/530992/82032358-5ea21800-969b-11ea-852d-4a7506388852.png)

In other words:

In the first step, you associate the formula with the system. But it's not locked! Association requires a highstate.
In the second step, you lock the system. And that does not require the highstate execution.

Author: ah ha, so the steps go like this:
* Apply the formula
* Apply highstate
* Configure formula (no need to apply highstate again)

??

Reviewer 2: It's a bit confusing. The first message should actually be "... apply the Highstate **if required**."
So the steps should be:
1. Enable the formula
2. Go the the formula and configure it
3. Save it and apply the Highstate only if the message asks you to do that. In the case of the "System Lock" formula it's not actually required.

Author: It doesn't hurt to apply the highstate, though, right? Why don't we just add it as an instruction regardless?

Reviewer: Yes, it does not hurt but it would be useless - IMHO we should not add that as an instruction.

//Something about requiring admin rights here maybe? --LKB 20200514

Some actions can still be completed on locked traditional clients, including remote commands, and automated patch updates.
To stop automated patch updates, navigate to the [guimenu]``System Details`` page for the client, and on the [guimenu]``Properties`` tab, uncheck [guimenu]``Auto Patch Update``.

== System Locks on Salt Clients

When a Salt client is locked, or put into blackout mode, no actions can be scheduled, Salt execution commands are disabled, and a yellow banner is displayed on the [guimenu]``System Details`` page.
In this mode, actions can be scheduled for the locked client using the {webui} or the API, but the actions will fail.

[NOTE]
====
The locking mechanism is not available for Salt SSH clients.
====

.Procedure: System Locking a Salt Client
. In the {productname} {webui}, navigate to the [guimenu]``System Details`` page for the client you want to lock.
. Navigate to the [guimenu]``Formulas`` tab, check the system lock formula, and click btn:[Save].
. Navigate to the menu:Formulas[System Lock] tab, check [guimenu]``Lock system``, and click btn:[Save].

> On this page, you can also specify specific Salt modules to enable them while the Salt client is locked.

Reviewer 3: unclear IMO.  what's "them"?  "specific" is superfluous (or not helpful enough).

> . If instructed to by the {webui}, apply the highstate.

Reviewer 3: ```suggestion
. If instructed by the {webui}, apply the highstate.
```

Author: OK, so this is interesting. In my original form, I've split the infinitive verb "to apply". Removing the "to" would make the sentence grammatically incorrect, but what we can do is remove the split, which would make it something like "If instructed _to apply_ the highstate by the {webui}, do it". Which, quite frankly, sounds awful. I'm reminded of Orwell's advice to "break any of these rules before doing anything truly barbarous" :joy: I think the only way I can make this sound any good without the split infinitive is to make it more repetitive, like: "If the {webui} instructs you to apply the highstate, apply the highstate." :woman_shrugging:
